### PR TITLE
feat(web,server)!: runtime log level

### PR DESCRIPTION
**Pull request by jrasm91**
> [!Warning]
> ### Breaking Change 
> `LOG_LEVEL` value of `simple` has been removed. 
> Use `log` instead (or remove the env entirely)

Change the log level at runtime:
- Log level options include: Fatal, Error, Warn, Log, Debug, and Verbose.
- The default log level is Log.
- Runtime logging is ignore when `LOG_LEVEL` is set at the environment variable level.
- An error is thrown why trying to change the log level from the UI (if environment variable is set)

image

image

### CODE REVIEW

**Review by bo0tzz — COMMENTED** [review]
This is nice! I do think we should keep the option to configure log level early via an env var, I can imagine things breaking hard enough that you can't even make it into the UI to set the log level. A config file would work there, but is relatively quite a bit more work to set up and might interfere with other troubleshooting.

File: server/src/infra/repositories/communication.repository.ts [hunk 1]
```diff
@@ -1,19 +1,60 @@
...
+  afterInit(server: Server) {
+    this.logger.log('Initialized websocket server');
+
+    for (const event of Object.values(ServerEvent)) {
+      server.on(event, async () => {
+        this.logger.debug(`Server event: ${event} (receive)`);
+        const callbacks = this.onServerEventCallbacks[event];
+        for (const callback of callbacks) {
+          await callback();
+        }
```
Comment (danieldietzler) [inline server/src/infra/repositories/communication.repository.ts:43]: Do we want to parallelize this by doing something like `await Promise.all(callbacks.map(callback => callback()))`

**Review by danieldietzler — APPROVED** [review]
Great job! Just some minor things... ;)
Also, I agree with Boet and I think we should actually keep the env variable for this

File: server/src/domain/system-config/system-config.service.ts [hunk 2]
```diff
@@ -23,14 +24,28 @@ import { SystemConfigCore, SystemConfigValidator } from './system-config.core';
...
+    this.core.config$.subscribe((config) => this.setLogLevel(config));
+  }
+
+  async init() {
+    const config = await this.core.getConfig();
+    await this.setLogLevel(config);
+  }
+
+  private async setLogLevel({ logging }: SystemConfig) {
+    const level = logging.enabled ? logging.level : false;
```
[the remark this answers is not in the document]
Comment (jrasm91) [reply server/src/domain/system-config/system-config.service.ts:46]: I believe that it correct

File: server/src/infra/logger.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    super(context);
+  }
+
+  isLevelEnabled(level: LogLevel) {
+    return isLogLevelEnabled(level, ImmichLogger.logLevels);
+  }
+
+  static setLogLevel(level: LogLevel | false): void {
+    ImmichLogger.logLevels = level === false ? [] : LOG_LEVELS.slice(LOG_LEVELS.indexOf(level));
+  }
```
[the remark this answers is not in the document]
Comment (jrasm91) [reply server/src/infra/logger.ts:20]: Maybe, this is for the existing nestjs logger works and I didn't bother reinventing the wheel here.

File: server/src/infra/repositories/communication.repository.ts (hunk 1 above)
Comment (jrasm91) [reply server/src/infra/repositories/communication.repository.ts:43]: We could but I think this is fine. Who knows if things depend on each other or stuff happening in a specific order. This is pretty predictable and performance is negligible here. Tbh, I think there is only even a single callback ever registered.

**Comment by jrasm91** [comment]
How should I resolve the two? Should I just ignore the runtime value while the environment variable is set?

File: server/src/infra/logger.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrasm91) [reply server/src/infra/logger.ts:20]: This is how the existing logger works right now (you can turn on or off each level independently). I will leave it for now, but agree we could update (simplify) it in the future.

**Review by bo0tzz — APPROVED** [review]

**Review by martabal — APPROVED** [review]

**Pull request merged**